src/radical/pilot/umgr/staging_input/default.py

```diff
@@ -338,12 +340,23 @@ def _handle_unit(self, unit, actionables):
             src = complete_url(src, src_context, self._log)
             tgt = complete_url(tgt, tgt_context, self._log)
 
-            if rpc.CREATE_PARENTS in flags:
-                copy_flags = rs.filesystem.CREATE_PARENTS
-            else:
-                copy_flags = 0
+            # Check if the src is a folder, if true
+            # add recursive flag if not already specified
+            if os.path.isdir(src.path):
+                if rs.filesystem.RECURSIVE not in flags:
+                    flags += [rs.filesystem.RECURSIVE]
 
-            saga_dir.copy(src, tgt, flags=copy_flags)
+            # Check if create parents flag is set,
+            # if not set it.
+            if rs.filesystem.CREATE_PARENTS not in flags:
+                flags += [rs.filesystem.CREATE_PARENTS]
+
+            if flags:
+                staging_flags = reduce(lambda x, y: x | y, flags)
+            else:
+                staging_flags = 0
+            
```

Comment:
Instead of reducing why not or the flags directly?

flags = 0
```
if rs.filesystem.RECURSIVE not in flags:
                     flags  = flags | rs.filesystem.RECURSIVE

if rs.filesystem.CREATE_PARENTS not in flags:
                 flags  = flags | rs.filesystem.CREATE_PARENTS
```

Reply:
True. But the user specifies the flags in the form of a list and not an integer. It looks like:
```
cud.input_staging = {'source': xyz, 'action': rp.TRANSFER, 'target': xyz, 'flags': [a,b,c]}
```
So we have to iterate through the list, performing what you suggest or use the reduce function at the end. Would you agree?

Reply:
my suggestion is to use instead of : 
```
flags += [rs.filesystem.RECURSIVE]

flags = flags |  rs.filesystem.RECURSIVE
```

So we will not have to reduce it at the end. Am I missing something?

Reply:
I don't think I understand what you are suggesting. 

``flags`` is a list. You are performing an OR with an int when you do ```flags = flags |  rs.filesystem.RECURSIVE``` so that will give you an error.

Second you are adding recursive flag to the list and then doing OR with it again.

I see two methods to do it:
```
>>> import saga.filesystem as rsf
>>> flags = [rsf.CREATE_PARENTS]
>>> flags += [rsf.RECURSIVE]
>>> flags
[64, 2]
>>> reduce(lambda x, y: x | y, flags)
66
```
or
```
>>> import saga.filesystem as rsf
>>> flags = [rsf.CREATE_PARENTS]
>>> flags += [rsf.RECURSIVE]
>>> flags
[64, 2]
>>> staging_flag = 0
>>> for f in flags:
...     staging_flag = staging_flag + f
... 
>>> staging_flag
66
```

Does this help? Is what you are suggesting similar to the second method here? If it is different maybe try the code out and paste the snippet here. That might help me understand.

Reply:
```
In [11]: import saga.filesystem as rsf

In [12]: staging_flag = 0

In [13]: staging_flag = staging_flag | rsf.CREATE_PARENTS

In [14]: staging_flag = staging_flag | rsf.RECURSIVE

In [15]: staging_flag
Out[15]: 66
```
that's it. If not flags are specified staging_flag will be 0. Let me know if it makes sense now

Reply:
What about other flags that the user might pass? (which we obtain in the form of a list)
see https://github.com/radical-cybertools/radical.pilot/blob/fb39871dc1da9bdd0c9896381f6c50190fb7b28d/src/radical/pilot/umgr/staging_input/default.py#L333

Reply:
I see, then we can't avoid the reduce.

Reply:
I am with George here though: If we use bitmask flags, then that should also apply to the API.  We can make this change backward compatible with:

```
if isinstance(sd['flags'], list):
    flags = sd['flags']
    sd['flags'] = 0
    for flag in flags: sd['flags'] |= flag
```

This can be applied once when we expand the staging directive, and then all other code paths can rely on flags being a bitmask.

This does not need to be part of the PR though.

Reply:
I am not sure I understand. Are you referring to the fact that current some flags are integers and some are strings? I agree that all flags should be changed to integers. But I don't understand how the lamda/reduce is different from what you are suggesting. Let me know.

Reply:
Sorry for not being clear.

I want to say that the API *should* use OR'ed integers, not a list.  We should change examples and documentation.  We can make lists work transparently by OR'ing the elements on the API layer.  If that is in place, you won't have to handle lists in the components.  But  all this does not need to be part of this RP.

Did that make more sense?

Reply:
I think I understand what you are suggesting, but not why. 

Just to be sure, what you are suggesting is:
* Within RP, ```cud['flags']``` will always be an integer as opposed to a list.
* At the API, the user needs to do the OR'ing between different flags or provides a list of flags that are converted into an integer and assigned to cud['flags'] so that the lower layers get an integer.
Correct me if I am wrong.

I am trying to understand the motivation behind changing something that works now.

Reply:
Yes: we should have OR'ed values everywhere (but can convert on the fly in the API if needed).  Motivation: that's how we do it in the other layers - and we really should not have two different ways to render flags, depending on what layer we use...

Reply:
Makes sense. Thanks!